### ISSUE TITLE
Proposal: async and load-balanced Muon/SOAP-style optimizers for TorchTitan

### ISSUE THREAD

**Issue by liangyuwang** [pytorch/torchtitan#3353]
Hi TorchTitan team,

I noticed several existing Muon / Scion / Dion related efforts in TorchTitan, including #1521, #1584, #1630, #2384, and #2494. However, from what I can see, a merged and directly usable fully distributed/asynchronous Muon-style optimizer path does not seem to be available on `main` yet.

We recently open-sourced **Megatron-Canzona**:
https://github.com/liangyuwang/Megatron-Canzona

Canzona is a unified, fully asynchronous, and load-balanced framework for distributed matrix-based optimizers such as Muon and SOAP. Its main focus is the system-level execution needed for large-scale pretraining, including:

- fully async and load-balanced optimizer pipeline across DP / TP (no redundant computation and memory)
- optimizer-agnostic support for arbitrary matrix-based optimizers, not limited to Muon or SOAP
- mixed Muon/SOAP + AdamW optimizer routing
- efficient handling of large/irregular matrices via parameter splitting, regrouping, adaptive buckets, and uneven collectives
- CUDA graph support for optimizer compute

We think TorchTitan could be a great place to develop a clean PyTorch-native version of this idea on top of FSDP2 / DTensor.

We would be happy to collaborate and contribute a PR, either as an experimental implementation or as a staged integration.

Would this be aligned with TorchTitan’s roadmap?

**Comment by tianyu-l** [comment]
Thanks for the proposal. How would you handle the situation where FSDP2 by default shards on dim-0 per parameter, but Muon require whole param gradient?

**Comment by liangyuwang** [comment]
Thanks for the question. We would not run Muon directly on each local FSDP2 shard, since that would change the algorithm. We also would not naively all-gather every full gradient to every rank.

Our initial plan is to generalize Canzona’s TP-ASC design to FSDP2-DP / TP sharding. FSDP2 dim-0 sharding can be treated as another matrix-fragmenting axis: after backward, different ranks collectively hold the full gradient of each 2D parameter as local DTensor shards.

The key point is that we do not process one parameter at a time. Instead, we first group the local gradients of multiple Muon-eligible parameters, or sub-parameters after splitting large matrices, into **micro gradient groups**. These groups are both the communication units and the load-balancing units.

For each micro gradient group, we would:

1. collect metadata for all parameters in the group, including global shape, local shard shape, DTensor placement, shard offsets, and target update layout;
2. pack the local gradient shards from multiple parameters into a temporary fused buffer;
3. use a single fused all-to-all, e.g. `all_to_all_single`, over the active FSDP2-DP / TP mesh to route shards to assigned host ranks;
4. reconstruct the full logical gradient for each matrix, or sub-matrix, only on its assigned host rank;
5. run Muon once on that reconstructed gradient;
6. slice the resulting update back into the original FSDP2 / TP shard layout;
7. pack update shards into a fused buffer and scatter them back with the reverse fused all-to-all;
8. let each original owner rank update its local DTensor shard.

This differs from a naive all-gather implementation in two ways:

- the full gradient is not materialized on every rank;
- Muon computation is not redundantly executed on every rank.

Instead, multiple parameters are fused into micro gradient groups, routed through all-to-all, and dynamically scheduled to host ranks for load balancing. This is the same idea as the TP-ASC pipeline in **Figure 2** of our paper, where micro gradient groups are used to saturate all-to-all bandwidth and reduce stragglers:
https://arxiv.org/pdf/2602.06079v1

The corresponding Megatron-Canzona design is documented here:
https://github.com/liangyuwang/Megatron-Canzona/blob/main/megatron/core/optimizer/matrix_based_optimizer/README.md#3-tensor-parallelism-async-compute-pipeline-load_balanced_tp_executorpy

For a first TorchTitan integration, we can keep the implementation simple by flattening the relevant FSDP2-DP / TP sharding axes into one optimizer communication group, while using DTensor metadata to reconstruct the correct per-parameter offsets and shard layouts.

Further communication optimizations can be added later, such as hierarchical DP/TP collectives, topology-aware host-rank scheduling, bucket reuse, or more specialized optimizer-state placement. The initial goal would be a clean and correct experimental path that avoids per-rank full-gradient materialization and avoids redundant Muon computation.

**Comment by tianyu-l** [comment]
Thanks. Another solution being proposed (and worked on) is drafted at https://github.com/pytorch/torchtitan/pull/2378/changes#diff-f967883f62f58914b3df1752452fdd8c2d151ae94550657c6a83e47cee966e9dR219-R236

The idea is that instead of using FSDP2 to shard each param on some dimension, we can put different params on different ranks, while each param stays unsharded (on some rank). The removes the need to all-gather / all-to-all on DP axis. One may still need comms across TP ranks.

Would love to hear what you think.

**Comment by liangyuwang** [comment]
Thanks for pointing this out. 

My previous answer was specifically under the assumption that we keep the default FSDP2 behavior, i.e. each parameter is sharded on dim-0 and the optimizer sees sharded per-parameter gradients. Under that setting, our proposal is to handle the full-gradient requirement of Muon at `optimizer.step()` time via Canzona-style micro gradient groups and fused all-to-all.

I agree that #2378 is another valid direction, but I would view it as a different parallelism/storage choice. If we put different whole parameters on different ranks, while each parameter stays unsharded on its owner rank, then this is closer to a DeepSpeed ZeRO-style / parameter-owner placement. It avoids DP-axis reconstruction for Muon because the owner already has the full parameter/gradient.

However, this requires changing the forward/backward parameter materialization and gradient communication path. In practice, it is no longer just an optimizer change; it becomes a different sharding/storage implementation for the training stack.

With the previous proposed FSDP2 + Canzona, we leave the FSDP2 fwd-bwd path unchanged and only add the extra reconstruction/routing logic during `optimizer.step()`. As shown in Figure 2 of our paper (https://arxiv.org/pdf/2602.06079v1), a large gradient group can be split into multiple micro-groups. Each micro-group contains gradients from multiple parameters or sub-parameters, and the micro-groups can be launched asynchronously. This allows the fused all-to-all communication of one group to overlap with the optimizer compute/update of another group, while also enabling load-balanced scheduling across host ranks.

For TorchTitan, our motivation is to provide a less intrusive and more PyTorch-native path: keep the default FSDP2 machinery for fwd/bwd, and add a Canzona optimizer-step pipeline for matrix-based optimizers. This may be easier to integrate and easier for users to adopt.

**Comment by liangyuwang** [comment]
<img width="1626" height="1044" alt="Image" src="https://github.com/user-attachments/assets/050b686c-c7ab-4916-a75c-4d9bd886a668" />

I can clarify my idea under the default FSDP2 assumption, where each parameter is sharded on dim-0 and the optimizer sees sharded per-parameter gradients.

For that setting, the idea is not to all-gather every full gradient to every rank, and also not to process parameters one by one. Instead, as illustrated in the sketch above, we can generalize Canzona’s TP-ASC pipeline to the FSDP2-DP axis.

A whole gradient group, e.g. `G1 ... G5`, is first split into multiple **micro gradient groups**. Each micro-group contains local gradient shards from multiple parameters, or sub-parameters after splitting large matrices.

For each matrix/sub-matrix, we maintain a stable load-balanced owner/host-rank assignment. The corresponding optimizer states, e.g. momentum or matrix-based preconditioner states, are fully and persistently kept on the assigned owner rank.

For each micro-group:

1. each FSDP2 rank packs its local gradient shards into a fused buffer;
2. a fused all-to-all routes these shards to the assigned owner/host ranks;
3. each owner rank reconstructs only the full gradients assigned to it, e.g. full `G1, G3` on one rank and full `G2` on another;
4. the owner rank runs Muon, or any matrix-based optimizer, once on the reconstructed full gradients using the optimizer states already kept on that rank;
5. the resulting update shards are scattered back with the reverse fused all-to-all;
6. each original FSDP2 rank updates its local parameter shards.

Multiple micro-groups can be launched asynchronously, so the gather/scatter communication of one micro-group can overlap with the optimizer compute/update of another micro-group.

This is the same principle as the TP-ASC design in Figure 2 of our paper:
https://arxiv.org/pdf/2602.06079v1

The key distinction from a naive FSDP2 all-gather is that the full gradient is not materialized on every rank, and Muon computation is not redundantly executed on every rank. The key distinction from a parameter-owner / ZeRO-style placement is that we keep the existing FSDP2 fwd/bwd path unchanged and add this pipeline only at `optimizer.step()` time.

For FSDP2 + TP, the same abstraction can be applied over the active sharding mesh: FSDP2-DP and TP are both treated as matrix-fragmenting axes, while DTensor metadata is used to recover the correct global offsets and shard layouts.

**Comment by tianyu-l** [comment]
cc @weifengpy @ailzhang

**Comment by liangyuwang** [comment]
We have implemented an FSDP version of Canzona here:
https://github.com/liangyuwang/FSDP-Canzona/

**Comment by weifengpy** [comment]
> We have implemented an FSDP version of Canzona here: https://github.com/liangyuwang/FSDP-Canzona/

thanks for the proposal! will take a deep look into the paper and codebase. we haven't investigated a lot into torch.distributed.optim a lot. it's a good timing to revisit the direction

**Comment by jadenmach2** [comment]
Following this thread with interest — the distributed-Muon design discussion (Canzona all-to-all vs the #2378
  param-owner placement) is all on NVIDIA so far. I've been validating Muon on AMD Instinct MI250 (gfx90a, ROCm 
  7.x, torch==2.10.0+rocm7.0) and would be glad to provide the ROCm reference/validation for whichever
  distributed design you land on.

  What I've verified end-to-end on MI250 so far:

  • FSDP2 (DP=2) ≡ single-GPU (AdamW): per-step loss diff 4.8e-7, final-weight diff 1.3e-5
  • Distributed Muon, Option A (all-gather grad → local NS → reshard) ≡ single-GPU Muon, fp32: loss diff
    4.8e-7, param diff 6.6e-7
  • Same in bf16 NS: loss diff 3.0e-4, param diff 2.6e-2 (at the bf16 NS precision floor)
  • DP=2 throughput scaling (Option A): 1.50× — per-GCD −25% from the all-gather + redundant NS on every rank
  • NS share of step time: ~68% on a 220M-param, dim-2048 model
  • Triton X@Xᵀ NS kernel (symmetry-exploiting), vs hipBLASLt: 1.03–1.08× on large transformer shapes

  Two takeaways relevant here: (1) the FP8 question is moot on CDNA2 (no FP8 hardware), which simplifies AMD
  validation; (2) the measured DP=2 = 1.50× and NS = 68% of step quantify exactly the inefficiency the
  async/load-balanced approach targets — so I can benchmark a candidate design against this Option A baseline on
   real AMD hardware.

  Happy to share the reproducible scripts and run AMD numbers for any design you'd like validated. Is ROCm validation something you'd want included as this moves toward a PR?
